src/routes/hydraauth.ts

```diff
@@ -0,0 +1,138 @@
+import {NextFunction, Request, Response} from 'express'
+import hydra from './../services/hydra.js';
+import config from '../config'
+import {sortFormFields} from '../translations'
+import {
+  AdminApi,
+  PublicApi,
+  FormField,
+  LoginRequest,
+  RegistrationRequest,
+} from '@oryd/kratos-client'
+import {IncomingMessage} from 'http'
+import url from 'url';
+import jd from 'jwt-decode';
+import {authInfo} from './home'
+import {protect} from '../index'
+
+// A simple express handler that shows the login / registration screen.
+// Argument "type" can either be "login" or "registration" and will
+// fetch the form data from ORY Kratos's Public API.
+const adminEndpoint = new AdminApi(config.kratos.admin)
+
+type UserRequest = Request & { user: any }
+
+export default (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  console.log("    --> hitting (auth/hydra/login ...")
+  const ai = authInfo(req as UserRequest)
+  console.log("    --> authInfo "+JSON.stringify(ai))
+  // The request is used to identify the login and registration request in ORY Kratos and return data like the csrf_token and so on.
+  const request = req.query.request
+  const currentLocation = `${req.protocol}://${req.headers.host}${req.url}`;
+
+  const query = url.parse(req.url, true).query;
+  // TODO FIGURE OUT HOW TO DO LOGIN AND REGISTRATION PAGES WITHOUT COOKIE FOR KEEPING THE CHALLENGE
+  let challenge = query.login_challenge || req.cookies.login_challenge;
+  if (challenge != null && challenge != "undefined") {
+    console.log("    --> Writing the login_challenge cookie from hydra: " + challenge);
+    res.cookie("login_challenge", challenge);
+  } else {
+    console.log("challenge exists: "+challenge)
+  } 
```

Comment:
this should be replaces with just parsing the login_challenge from the redirect_to response from kratos and putting it as a query param on the registration link

Reply:
I agree

Reply:
Ah, for registration we actually thought that almost no-one redirects from registration to OAuth2 Callback. Neither Facebook, Google, ... do that really. So we wanted to save some complexity here and just ignore registration. What do you think?

Reply:
I think this is not only registration-specific. We need the challenge in most if not all cases 
e.g. when calling the acceptLoginRequest:
https://github.com/k9ert/kratos-selfservice-ui-node/blob/master/src/routes/hydraauth.ts#L95

But in that case the kratos-request which could give us the challenge is long gone.
So imho there is currently no way around using that Cookie, right?!

Reply:
No cookie for the login_challenge needed anymore!
This has been resolved by using the default-login endpoint. The returnTo-Url is only relevant at the very end of the flow and it therefore beautifully storing the the login-challenge for us while we're doing the login via the default endpoint which is no longer aware that this is running within a hydra-flow and so we also don't need it for "state-management".